File: engine/lab/tests/test_stepper.py
"""Tests for the step controllers (DESIGN §5.3)."""

import pytest

from cloze_lab.scheduler.policies import StepContext
from cloze_lab.scheduler.stepper import (
    AdaptiveStepper,
    FixedStepper,
    StepController,
    StepOutcome,
)


def outcome(remaining: int, step: int = 0, committed: int = 1) -> StepOutcome:
    return StepOutcome(step=step, n_committed=committed, n_masked_after=remaining)


class TestFixedStepper:
    def test_satisfies_protocol(self) -> None:
        assert isinstance(FixedStepper(4), StepController)

    def test_cap_and_context(self) -> None:
        s = FixedStepper(4)
        assert s.steps_cap == 4
        ctx = s.context(1)
        assert ctx == StepContext(step=1, steps_total=4)  # quota mode keeps its budget

    def test_continue_until_drained(self) -> None:
        s = FixedStepper(4)
        assert s.should_continue(outcome(remaining=3)) is True
        assert s.should_continue(outcome(remaining=0)) is False

    @pytest.mark.parametrize("bad", [0, -1])
    def test_rejects_nonpositive_steps(self, bad: int) -> None:
        with pytest.raises(ValueError, match="steps must be"):
            FixedStepper(bad)


class TestAdaptiveStepper:
    def test_satisfies_protocol(self) -> None:
        assert isinstance(AdaptiveStepper(8), StepController)

    def test_cap_and_context(self) -> None:
        s = AdaptiveStepper(8)
        assert s.steps_cap == 8
        # No fixed budget under adaptive -> steps_total None (quota mode would raise).
        assert s.context(3) == StepContext(step=3, steps_total=None)

    def test_continue_until_drained(self) -> None:
        s = AdaptiveStepper(8)
        assert s.should_continue(outcome(remaining=2)) is True
        assert s.should_continue(outcome(remaining=0)) is False

    @pytest.mark.parametrize("bad", [0, -3])
    def test_rejects_nonpositive_tmax(self, bad: int) -> None:
        with pytest.raises(ValueError, match="t_max must be"):
            AdaptiveStepper(bad)


def test_fixed_context_matches_legacy_inline_construction() -> None:
    # The load-bearing compatibility line: FixedStepper(T).context(step) must equal
    # the StepContext the pre-stepper loop built inline (StepContext(step, steps_total=T)).
    for steps in (1, 3, 8):
        s = FixedStepper(steps)
        for step in range(steps):
            assert s.context(step) == StepContext(step=step, steps_total=steps)

File: engine/core/tests/test_stepper.cpp
// test_stepper.cpp — checks for the C++ stepper port, mirroring test_stepper.py.
#include "cloze/stepper.hpp"

#include <cassert>
#include <cstdio>
#include <stdexcept>

using namespace cloze;

int main() {
    // fixed(T): advertises the budget so quota mode keeps it; caps at T.
    {
        auto s = Stepper::fixed(4);
        assert(s.steps_cap() == 4);
        assert(s.context(0).steps_total == 4);
        assert(s.context(0).steps_remaining() == 4);
        assert(s.context(3).steps_remaining() == 1);
        assert(s.should_continue(StepOutcome{0, 1, 2}) == true);   // masks remain
        assert(s.should_continue(StepOutcome{3, 2, 0}) == false);  // drained
    }
    // adaptive(T_max): no step budget (steps_total = -1), caps at t_max.
    {
        auto s = Stepper::adaptive(8);
        assert(s.steps_cap() == 8);
        assert(s.context(0).steps_total == -1);
        assert(s.context(0).steps_remaining() == -1);
        assert(s.should_continue(StepOutcome{0, 1, 3}) == true);
    }
    // validation: budgets must be >= 1.
    {
        bool threw = false;
        try { Stepper::fixed(0); } catch (const std::invalid_argument&) { threw = true; }
        assert(threw);
        threw = false;
        try { Stepper::adaptive(0); } catch (const std::invalid_argument&) { threw = true; }
        assert(threw);
    }
    std::printf("test_stepper: all assertions passed\n");
    return 0;
}
